astropy/io/ascii/cparser.pyx

```diff
@@ -386,7 +386,12 @@ cdef class CParser:
             data_end = max(self.data_end - self.data_start, 0) # read nothing if data_end < 0
 
         if tokenize(self.tokenizer, data_end, 0, <int>len(self.names)) != 0:
-            self.raise_error("an error occurred while parsing table data")
+            if self.tokenizer.code in (NOT_ENOUGH_COLS, TOO_MANY_COLS):
+                raise core.InconsistentTableError("Number of header columns " +
```

Comment:
Raising consistent exceptions on inconsistent tables between the slow and fast readers.